lec08-stability.Rmd

```diff
@@ -1,4 +1,576 @@
+points <- tmax/dt # the number of data points in the simulation - add 1 so that we can start at t=0
+
+# vectors to store values of X, Y, and t at each timestep:
+X_vector <- numeric(points) # prey population size
+Y_vector <- numeric(points) # predator population size
+t_vector <- seq(0, tmax - dt, by = dt) # time vector
+
+# initial condition
+X0 <- 2
+Y0 <- 2
+X_vector[1] <- X0
+Y_vector[1] <- Y0
+
+X <- X0 # initialize variable X
+Y <- Y0 # initialize variable Y
+```
+
+This is the right way to do it.
+
+```{r}
+for (i in 2:points) {
+  # start at 2 because the initial state is at position 1
+  
+  # first calculate BOTH dX and dY
+  dX <- dX_dt(X = X, Y = Y, a = a, b = b) * dt
+  dY <- dY_dt(X = X, Y = Y, c = c, d = d) * dt
+  
+  # now update BOTH X and Y
+  X <- X + dX 
+  Y <- Y + dY
+  X_vector[i] <- X
+  Y_vector[i] <- Y
+}
+```
+
+This is the wrong way.
+
+```{r}
+for (i in 2:points) {
+  # start at 2 because the initial state is at position 1
+  
+  # calculate dX and then update X (wrong)
+  dX <- dX_dt(X = X, Y = Y, a = a, b = b) * dt
+  X < X + dX
+  
+  # calculate dY and then update Y - this will use the new value of X instead of the old one.
+  dY <- dY_dt(X = X, Y = Y, c = c, d = d) * dt
+  Y <- Y + dY
+  
+  X_vector[i] <- X
+  Y_vector[i] <- Y
+}
+```
+
+## Phase portraits in two dimensions
+
+#### Challenge
+
+What were the axes of the one-dimensional phase portraits we drew? What would the axes
+of a phase portrait for the predator-prey system above be?
+
+```{r, eval=FALSE, echo=FALSE}
+# Challenge solution
+
+# The axis was N in the 1D version. 
+# For a 2D system with variables x and y, the two axes would be x and y.
+```
+
+Continuing with the predator-prey example: instead of plotting our numerical trajectory as $X$ and 
+$Y$ vs. $t$, we'll plot $Y$ vs. $X$: predator population size vs. prey population size. 
+
+```{r}
+ggplot(result) +
+  geom_point(aes(x = X, y = Y)) # geom_line connects the dots strangely
+```
+
+A phase portrait in two dimensions has two axes: one for each of the variables in the system.
+Here, the two variables are $X$ and $Y$, prey and predators. We want to include the same 
+information in this 2D phase portrait as in the 1D portrait: at minimum, we want to mark 
+the fixed points and draw a **vector field**: arrows indicating in which direction the rate 
+of change is for different values of $X$ and $Y$. 
+
+We can also add other information, and in the plot above we haven't yet drawn arrows or
+fixed points, but we have drawn a **trajectory** - the dots trace out what happens to $X$
+and $Y$ over time starting from $X = 2$, $Y = 2$. Time is no longer an axis in the plot,
+but it's still there, hidden in the path that the populations take. 
+
+Comparing these two ways of plotting a trajectory, we can see that oscillations in time
+look like circular motion in the phase portrait. The GIF below illustrates this.
+
+```{r}
+library(animation)
```

Comment:
I'm not sure why there is an error on your local machine, but the Travis error is because you are using this package without including it in the `DESCRIPTION` under `Imports:` (which you can easily add from the console via `devtools::use_package("animation")`)

Reply:
On my machine it's because ImageMagick isn't working. The space in the folder name `Program Files` is f***ing it up, and I'm not too invested in fixing it. 

I might change that block to not evaluate anyway because I usually upload the GIF separately after it's created.